packages/block-editor/src/utils/dom.js

```diff
@@ -57,3 +57,110 @@ export function getBlockClientId( node ) {
 
 	return blockNode.id.slice( 'block-'.length );
 }
+
+/**
+ * Calculates the union of two rectangles, and optionally constrains this union within a containerRect's
+ * left and right values.
+ * The function returns a new DOMRect object representing this union.
+ *
+ * @param {DOMRect}          rect1         First rectangle.
+ * @param {DOMRect}          rect2         Second rectangle.
+ * @param {DOMRectReadOnly?} containerRect An optional container rectangle. The union will be clipped to this rectangle.
+ * @return {DOMRect} Union of the two rectangles.
+ */
+export function rectUnion( rect1, rect2, containerRect ) {
+	let left = Math.min( rect1.left, rect2.left );
+	let right = Math.max( rect1.right, rect2.right );
+	const bottom = Math.max( rect1.bottom, rect2.bottom );
+	const top = Math.min( rect1.top, rect2.top );
+
+	/*
+	 * To calculate visible bounds using rectUnion, take into account the outer
+	 * horizontal limits of the container in which an element is supposed to be "visible".
+	 * For example, if an element is positioned -10px to the left of the window x value (0),
+	 * this function discounts the negative overhang because it's not visible and
+	 * therefore not to be counted in the visibility calculations.
+	 * Top and bottom values are not accounted for to accommodate vertical scroll.
+	 */
+	if ( containerRect ) {
+		left = Math.max( left, containerRect.left );
+		right = Math.min( right, containerRect.right );
+	}
+
+	return new window.DOMRect( left, top, right - left, bottom - top );
+}
+
+/**
+ * Returns whether an element is visible.
+ *
+ * @param {Element} element Element.
+ * @return {boolean} Whether the element is visible.
+ */
+function isElementVisible( element ) {
+	const viewport = element.ownerDocument.defaultView;
+	if ( ! viewport ) {
+		return false;
+	}
+
+	// Check for <VisuallyHidden> component.
+	if ( element.classList.contains( 'components-visually-hidden' ) ) {
+		return false;
+	}
+
+	const bounds = element.getBoundingClientRect();
```

Comment:
Do we still need this if we're using `checkVisibility`?

Reply:
Good question.

I believe so because this tests for elements that would otherwise be visible if not for the `0` dimensions.

E.g., for `<div id="test" style="width:0;height:0;">Test</div>`, `checkVisibility( ...options )` returns `true`.

https://drafts.csswg.org/cssom-view-1/#dom-element-checkvisibility

Reply:
I'm not sure this is supported in all the browsers that we support. I keep getting errors when using Safari.

Reply:
Same problem here

Reply:
Based on the [browserslist config](https://github.com/WordPress/gutenberg/blob/trunk/packages/browserslist-config/index.js) we support, we support the [last 2 versions of Safari](https://browsersl.ist/#q=%3E+1%25%0Alast+1+Android+versions%0Alast+1+ChromeAndroid+versions%0Alast+2+Chrome+versions%0Alast+2+Firefox+versions%0Alast+2+Safari+versions%0Alast+2+iOS+versions%0Alast+2+Edge+versions%0Alast+2+Opera+versions) (17.5, 17.4). Both versions support `checkVisibility` according to [caniuse](https://caniuse.com/mdn-api_element_checkvisibility).

Maybe the config we use is too loose if this is unexpected?

Reply:
Safari only updates with MacOS updates for me. (stable MacOS) and I'm still on 17.3 for this reason. I think we might want to update our Safari browser list config to 3 or 4 versions because of this.

Reply:
Good catch — yeah, I think we should handle calling this optionally. Even if we officially support the latest two Safari versions, we shouldn't be in a situation where we're breaking the editor in older Safari versions. Here's what I'm seeing when I select a block in the post editor in Safari 16.4.1:

<img width="1517" alt="image" src="https://github.com/user-attachments/assets/d2a74f3e-0b20-4e66-a992-bd59a7fcde36">

I've put up a quick fix PR here: https://github.com/WordPress/gutenberg/pull/65069